### PR TITLE
Add fatal error to IAR parser

### PR DESCRIPTION
**Pull request by kvdz**
Add the short error, Fatal error detection and better detection via the regular expression.

- update test cases too have one test at the time.
- update the resource file so all cases can be tested via the unit test

### CODE REVIEW

File: src/main/java/hudson/plugins/warnings/parser/IarParser.java [hunk 1]
```diff
@@ -14,27 +14,30 @@
...
+    private static final String IAR_WARNING_PATTERN = 
+"((\\[exec\\] )?(.*)\\((\\d+)\\)?.*)?(Fatal [Ee]rror|Remark|Warning)\\[(\\w+)\\]: (.*(\\\".*(c|h)\\\")|.*)";
     /**
      * Creates a new instance of {@link IarParser}.
      */
     public IarParser() {
         super(Messages._Warnings_iar_ParserName(),
                 Messages._Warnings_iar_LinkName(),
                 Messages._Warnings_iar_TrendName(),
-                IAR_WARNING_PATTERN, true);
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/IarParser.java]: If you remove the true, then the method isLineInteresting is **not** invoked anymore!

File: src/main/java/hudson/plugins/warnings/parser/IarParser.java [hunk 2]
```diff
@@ -45,26 +48,30 @@ protected String getId() {
...
+           
+        priority = determinePriority(matcher.group(GROUP_NUMBER));
+        return composeWarning(matcher, priority);
+    }
+           
+    private Warning composeWarning(final Matcher matcher, final Priority priority) {
+        String message = matcher.group(7);
+        
+        if( matcher.group(3) == null ) {
+            // createWarning( filename, line number, error number (Pe177), message, priority )
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/IarParser.java]: Remove `// comments`

File: src/main/java/hudson/plugins/warnings/parser/IarParser.java [hunk 3]
```diff
@@ -45,26 +48,30 @@ protected String getId() {
...
+            return createWarning(matcher.group(8), 0, matcher.group(6), message, priority);
         }
-        String message = normalizeWhitespaceInMessage(matcher.group(5));
-        return createWarning(matcher.group(1), getLineNumber(matcher.group(2)), matcher.group(4), message, priority);
+        // createWarning( filename, line number, error number (Pe177), message, priority )
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/IarParser.java]: Remove `// comments`

File: src/test/java/hudson/plugins/warnings/parser/IarParserTest.java [hunk 4]
```diff
@@ -20,55 +20,146 @@
...
+     /**
+     * Parses a file and check error number 2
+     *
+     * @throws IOException
+     *      if the file could not be read
+     * @see <a href="http://issues.jenkins-ci.org/browse/JENKINS-8823">Issue 8823</a>
+     */
+   @Test
+    public void IAR_error2() throws IOException {
+        Collection<FileAnnotation> warnings = new IarParser().parse(openFile("issue8823.txt"));
```
Comment (uhafner) [inline src/test/java/hudson/plugins/warnings/parser/IarParserTest.java]: Can you wrap that into a method with an integer parameter for the warning number to return?
```
Collection<FileAnnotation> warnings = new IarParser().parse(openFile("issue8823.txt"));
Iterator<FileAnnotation> iterator = warnings.iterator();
       
For loop: iterator.next();
 
return  iterator.next();
```

Comment (kvdz) [inline src/test/java/hudson/plugins/warnings/parser/IarParserTest.java]: Good idea!
something like this?
`protected FileAnnotation getErrorNumber( final int number)
    {
        Collection<FileAnnotation> warnings = new IarParser().parse(openFile("issue8823.txt"));
        Iterator<FileAnnotation> iterator = warnings.iterator();
        FileAnnotation annotation = iterator.next();

        for(int i = 1; i < number; i++)
        {
            annotation = iterator.next();
        }
        
        return annotation;
    }`

**Comment by kvdz** [comment]
@uhafner  i have incorporated your review

**Comment by uhafner** [comment]
Can you please have a look at JENKINS-48051? Seems that this change drops the performance...

**Pull request merged**